### PR TITLE
closes #25 | user sign up

### LINKED ISSUE #25: Add user sign up capabilities
Tailors should signup via the `/tailors/signup` route

Customers should signup via the `/customers/signup` route

For Customers
*Request*

```json
{
  "firstname": "xxxx",
  "lastname": "xxxxx",
  "phone": "xxxxx",
  "email": "xxxxx",
  "password": "xxxx"
}
```

For Tailors
*Request*

```json
{
  "firstname": "xxxx",
  "lastname": "xxxxx",
  "phone": "xxxxx",
  "email": "xxxxx",
  "location": "xxxxx",
  "password": "xxxx",
  "outfitTypes": ["xxxxx","xxxx"]
}
```

*Response*

```json
{
  "jwt": "xxxxx"
}
```

**NOTE**: the `payload ` for the `jwt` is to be the user `_id`.

### PR DESCRIPTION
**Pull request by bolorundurowb**
*Description of Task*
- Implement sign up for tailors and customers

### CODE REVIEW

File: src/server/controllers/Authentication.js [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        lastName: body.lastName,
+        email: body.email,
+        password: body.password,
+        phone: body.phone,
+        role: type
+      };
+
+      if (req.file) {
+        Authentication.uploadImage(req.file)
+          .then((url) => {
```
Comment (Saint-lawal) [inline src/server/controllers/Authentication.js]: So what happens when a users picture isn't saved for whatever reason?

File: src/server/middleware/Authentication.js [hunk 2]
```diff
@@ -44,6 +46,32 @@ export default class AuthenticationMiddleware {
...
+              });
+            } else if (user.role === 2) {
+              Tailor.findOne({ uid: user._id }, (err, tailor) => {
+                if (err) {
+                  res.status(500).send({
+                    message: 'An error occurred when retrieving the tailor.',
+                  });
+                } else {
+                  req.user = user;
+                  req.tailor = tailor;
```
Comment (Saint-lawal) [inline src/server/middleware/Authentication.js]: We don't need to have 2 different objects, we could just have req.user and then req.user.info can fold the tailor information

File: src/server/middleware/Authentication.js [hunk 3]
```diff
@@ -44,6 +46,32 @@ export default class AuthenticationMiddleware {
...
               });
+            } else if (user.role === 1) {
+              Customer.findOne({ uid: user._id }, (err, customer) => {
+                if (err) {
+                  res.status(500).send({
+                    message: 'An error occurred when retrieving the customer.',
+                  });
+                } else {
+                  req.user = user;
+                  req.customer = customer;
```
Comment (Saint-lawal) [inline src/server/middleware/Authentication.js]: Same as below

File: src/server/controllers/Authentication.js (hunk 1 above)
Comment (bolorundurowb) [inline src/server/controllers/Authentication.js]: @Saint-lawal `cloudinary` doesn't provide an error callback. So I can't check for that

File: src/server/middleware/Authentication.js (hunk 2 above)
Comment (bolorundurowb) [inline src/server/middleware/Authentication.js]: I initially wanted it to be `req.customer/tailor` and `req.customer/tailor.user` since that preserves the hierarchy but I considered the admins and super users who do not follow that hierarchy. So the reason why I did this is so that if you are dealing with a `customer` you know where to look and same with `tailors`

File: src/server/controllers/Authentication.js (hunk 1 above)
Comment (Saint-lawal) [inline src/server/controllers/Authentication.js]: Makes sense.

File: src/server/middleware/Authentication.js (hunk 2 above)
Comment (Saint-lawal) [inline src/server/middleware/Authentication.js]: Alright 👍

File: src/server/controllers/Authentication.js [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+          uid: user._id,
+          dateOfBirth: body.dateOfBirth,
+          address: body.address,
+          location: body.location,
+          bankDetails: body.bankDetails,
+          referrerName: body.referrerName,
+          outfitTypeId: body.outfitTypeId
+        });
+
+        tailor.save((err, _tailor) => {
```
Comment (adetola-ralph) [inline src/server/controllers/Authentication.js]: You can use a promise `then` here

File: src/server/controllers/Authentication.js [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+      expiresIn: '24h'
+    });
+  }
+
+  /**
+   * Uploads an image to cloudinary
+   * @param {Object} file
+   * @return {Promise<String>} the picture url
+   */
+  static uploadImage(file) {
```
Comment (adetola-ralph) [inline src/server/controllers/Authentication.js]: Why is this in authentication controller?

File: src/server/controllers/Authentication.js [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+      }
+    });
+  }
+
+  /**
+   * Commits a user model instance to the database
+   * @param {Object} user
+   * @returns {Promise<Object>} the saved user
+   */
+  static commitUserToDb(user) {
```
Comment (adetola-ralph) [inline src/server/controllers/Authentication.js]: this is redundant, mongoose > v4 supports promises from the save function

File: src/server/middleware/Authentication.js [hunk 7]
```diff
@@ -35,7 +37,7 @@ export default class AuthenticationMiddleware {
             message: 'Invalid token',
           });
         } else {
-          User.findById(decoded._id, (err, user) => {
```
Comment (adetola-ralph) [inline src/server/middleware/Authentication.js]: Why the change here, I thought mongoose returns the field as _id and not id?

File: src/server/middleware/Validations.js [hunk 8]
```diff
@@ -23,6 +23,81 @@ class Validations {
...
+
+    if (!body.phone) {
+      res.status(400).send({ message: 'A phone number is required.' });
+    } else if (!Validations.isMobileValid(body.phone)) {
+      res.status(400).send({ message: 'The phone number provided is invalid.' });
+    } else if (!body.password) {
+      res.status(400).send({ message: 'A password is required.' });
+    } else if (body.password.length <= 6) {
+      res.status(400).send({ message: 'Passwords must be longer than 6 characters.' });
+    } else if (body.email) {
```
Comment (adetola-ralph) [inline src/server/middleware/Validations.js]: I think email is necessary, cause we're building a web client for them to signup and signin

File: src/server/middleware/Authentication.js (hunk 7 above)
Comment (bolorundurowb) [inline src/server/middleware/Authentication.js]: @adetola-ralph I create the token with the field as `id` since that is cleaner than `_id`

File: src/server/middleware/Validations.js (hunk 8 above)
Comment (bolorundurowb) [inline src/server/middleware/Validations.js]: @adetola-ralph They can signup and signn with their Phone numbers

File: src/server/controllers/Authentication.js (hunk 6 above)
Comment (bolorundurowb) [inline src/server/controllers/Authentication.js]: I set that there to reuse for two controller methods @adetola-ralph

File: src/server/controllers/Authentication.js (hunk 5 above)
Comment (bolorundurowb) [inline src/server/controllers/Authentication.js]: Because the Authentication methods use it and none others

File: src/server/controllers/Authentication.js (hunk 4 above)
Comment (bolorundurowb) [inline src/server/controllers/Authentication.js]: yeah, but using the callback reduces the nesting and the promise is still less than ideal especially when there are errors
